src/cast_core/src/RollingCollections.cpp

```diff
@@ -269,50 +295,127 @@ struct OplogTailer : public RunOperation {
-                // Do nothing.
+    bool isRollingOplogEntry(const bsoncxx::document::view &doc, long *rollingMillis)
+    {
+        if (doc["op"].get_utf8().value.to_string() == "c") {
+            auto object = doc["o"].get_document().value;
+            auto it = object.find("create");
+            if (it != object.end()) {
+                auto collectionName = object["create"].get_utf8().value.to_string();
+                if (collectionName.substr(0, 2) == "r_") {
+                    // It's a collection we care about,
+                    // get the embedded millisecond time.
+                    std::vector<std::string> timeSplit;
+                    boost::algorithm::split(timeSplit, collectionName, boost::is_any_of("_"));
+                    *rollingMillis = stol(timeSplit[2]);
+                    return (true);
+                }
+            }
+        }
+        return (false);
+    }
+
+    // When we're still catching up, our lag times will be getting better
+    // and better. As a simple determinant, we'll say we're done catching
+    // up if we haven't a new "best" lag time within the last 5 seen.
+    bool caughtUp(uint64_t lagNanoseconds) {
+        if (!_caughtUp) {
+            if (lagNanoseconds < _catchUpBestLag) {
+                _catchUpBestLag = lagNanoseconds;
+                _catchUpBestWhen = 0;
+            } else if (++_catchUpBestWhen > 5) {
+                _caughtUp = true;
+                BOOST_LOG_TRIVIAL(info) << "Oplog tailer: caught up";
             }
-            _firstLoop = false;
         }
-        for (auto&& doc : _cursor.value()) {
-            if (doc["op"].get_utf8().value.to_string() == "c") {
-                auto object = doc["o"].get_document().value;
-                auto it = object.find("create");
-                if (it != object.end()) {
-                    auto collectionName = object["create"].get_utf8().value.to_string();
-                    if (collectionName.length() > 2 && collectionName[0] == 'r' &&
-                        collectionName[1] == '_') {
-                        // Get the time as soon as we know we its a collection we care about.
-                        auto now = metrics::clock::now();
-                        std::vector<std::string> timeSplit;
-                        boost::algorithm::split(timeSplit, collectionName, boost::is_any_of("_"));
-                        auto started = metrics::clock::time_point{
-                            std::chrono::duration_cast<metrics::clock::time_point::duration>(
-                                std::chrono::milliseconds{std::stol(timeSplit[2])})};
-                        _oplogLagOperation.report(
-                            now,
-                            std::chrono::duration_cast<std::chrono::microseconds>(now - started),
-                            metrics::OutcomeType::kSuccess);
+        return (_caughtUp);
+    }
+
+    // Called when we see a create of a rolling collection.
```

Comment:
Possibly better to use block comments here?

Reply:
I think this is a C++-ism.  Looking at the genny source code, it looks like it primarily uses // for block commenting, and /** ... */ style for "javadoc style" function comments.  @rtimmons, any opinion on comment styles?

Reply:
no opinion here - double-asterisk is for javadoc-style as you indicated. I personally don' like 

```
/*
foo
*/
```

style comments but that's totally personal-preference.

Reply:
I'm going to leave multiline comments using '//' to be closer with other genny/mongodb c++ style I've seen.